### PR TITLE
Extended pairwise addition instructions

### PR DESCRIPTION
**Pull request by Maratyszcza**
Introduction
=========

This PR introduce extended pairwise addition operation that computes extended sums within adjacent pairs of lanes of a SIMD vector, and produce a SIMD vector with twice fewer of twice wider lanes. This operation naturally arise when doing partial reductions in fixed-point algorithms, nicely maps to a single instruction on ARM, and can be simulated with just 1-4 instructions on SSE4+ x86.

Applications
=========

- [libvpx video codec](https://github.com/webmproject/libvpx/blob/a5d499e16570d00d5e1348b1c7977ced7af3670f/vp8/encoder/arm/neon/denoising_neon.c#L127-L132)
- [x264 video codec](https://github.com/mirror/x264/blob/33f9e1474613f59392be5ab6a7e7abf60fa63622/common/aarch64/pixel-a.S#L485-L486)
- [TensorFlow Lite neural network framework](https://github.com/aiswaryarajm/tensorflow/blob/9cef7b78b0e9f20cd828111477e35b619d59617d/tensorflow/lite/kernels/internal/optimized/depthwiseconv_3x3_filter_common.h#L148-L149)
- [MegEngine neural network framework](https://github.com/MegEngine/MegEngine/blob/19568f67e9e80ae729272f56f19b739a72e5b889/dnn/src/arm_common/conv_bias/int8x8x16/conv_stride2.cpp#L81)
- [QNNPACK quantized neural network library](https://github.com/pytorch/QNNPACK/blob/639ea0b0b0ddaaf2122f52c1e472a65941d57372/src/q8gemm/4x-sumrows-neon.c#L59-L62)
- [Ruy matrix-matrix multiplication library](https://github.com/google/ruy/blob/59c2de870307b80a59795c3768c954d80159c838/ruy/pack_arm.cc#L83-L97)
- [3PXNet neural network library](https://github.com/nanocad-lab/3pxnet-inference/blob/9932694ca8c4b7964fbc0f1cfeabc7c57e3ef12f/xnor/xnor_fc.c#L395-L399)

Mapping to Common Instruction Sets
===========================

This section illustrates how the new WebAssembly instructions can be lowered on common instruction sets. However, these patterns are provided only for convenience, compliant WebAssembly implementations do not have to follow the same code generation patterns.

x86/x86-64 processors with XOP instruction set
--------------------------------------------------

- **i16x8.extadd_pairwise_i8x16_s**
  - `y = i16x8.extadd_pairwise_i8x16_s(x)` is lowered to `VPHADDBW xmm_y, xmm_x`
- **i16x8.extadd_pairwise_i8x16_u**
  - `y = i16x8.extadd_pairwise_i8x16_u(x)` is lowered to `VPHADDUBW xmm_y, xmm_x`
- **i32x4.extadd_pairwise_i16x8_s**
  - `y = i32x4.extadd_pairwise_i16x8_s(x)` is lowered to `VPHADDWD xmm_y, xmm_x`
- **i32x4.extadd_pairwise_i16x8_u**
  - `y = i32x4.extadd_pairwise_i16x8_u(x)` is lowered to `VPHADDUWD xmm_y, xmm_x`

x86/x86-64 processors with AVX instruction set
--------------------------------------------------

- **i16x8.extadd_pairwise_i8x16_s**
  - `y = i16x8.extadd_pairwise_i8x16_s(x)` (`y` is **NOT** `x`) is lowered to:
    - `VMOVDQA xmm_y, [wasm_i8x16_splat(1)]`
    - `VPMADDUBSW xmm_y, xmm_y, xmm_x`
- **i16x8.extadd_pairwise_i8x16_u**
  - `y = i16x8.extadd_pairwise_i8x16_u(x)` is lowered to `VPMADDUBSW xmm_y, xmm_x, [wasm_i8x16_splat(1)]`
- **i32x4.extadd_pairwise_i16x8_s**
  - `y = i32x4.extadd_pairwise_i16x8_s(x)` is lowered to `VPMADDWD xmm_y, xmm_x, [wasm_i16x8_splat(1)]`
- **i32x4.extadd_pairwise_i16x8_u**
  - `y = i32x4.extadd_pairwise_i16x8_u(x)` (`y` is **not** `x`) is lowered to:
    - `VPSRLD xmm_y, 16`
    - `VPBLENDW xmm_tmp, xmm_x, xmm_y, 0xAA`
    - `VPADDD xmm_y, xmm_y, xmm_tmp`

x86/x86-64 processors with SSSE3 instruction set
--------------------------------------------------

- **i16x8.extadd_pairwise_i8x16_s**
  - `y = i16x8.extadd_pairwise_i8x16_s(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVDQA xmm_y, [wasm_i8x16_splat(1)]`
    - `PMADDUBSW xmm_y, xmm_x`
- **i16x8.extadd_pairwise_i8x16_u**
  - `y = i16x8.extadd_pairwise_i8x16_u(x)` is lowered to:
    - `MOVDQA xmm_y, xmm_x`
    - `PMADDUBSW xmm_y, [wasm_i8x16_splat(1)]`

x86/x86-64 processors with SSE2 instruction set
--------------------------------------------------

- **i16x8.extadd_pairwise_i8x16_s**
  - `y = i16x8.extadd_pairwise_i8x16_s(x)` (`y` is **NOT** `x`) is lowered to:
    - `MOVDQA xmm_tmp, xmm_x`
    - `MOVDQA xmm_y, xmm_x`
    - `PSLLW xmm_tmp, 8`
    - `PSRAW xmm_y, 8`
    - `PSRAW xmm_tmp, 8`
    - `PADDW xmm_y, xmm_tmp`
- **i16x8.extadd_pairwise_i8x16_u**
  - `y = i16x8.extadd_pairwise_i8x16_u(x)` is lowered to:
    - `MOVDQA xmm_tmp, [wasm_i16x8_splat(0x00FF)]`
    - `MOVDQA xmm_y, xmm_x`
    - `PAND xmm_tmp, xmm_x`
    - `PSRLW xmm_y, 8`
    - `PADDW xmm_y, xmm_tmp`
- **i32x4.extadd_pairwise_i16x8_s**
  - `y = i32x4.extadd_pairwise_i16x8_s(x)` is lowered to:
    - `MOVDQA xmm_y, xmm_x`
    - `PMADDWD xmm_y, [wasm_i16x8_splat(1)]`
- **i32x4.extadd_pairwise_i16x8_u**
  - `y = i32x4.extadd_pairwise_i16x8_u(x)` is lowered to:
    - `MOVDQA xmm_y, xmm_x`
    - `PXOR xmm_y, [wasm_i16x8_splat(0x8000)]`
    - `PMADDWD xmm_y, [wasm_i16x8_splat(1)]`
    - `PADDD xmm_y, [wasm_i32x4_splat(0x00010000)]`

ARM64 processors
--------------------------------------------------

- **i16x8.extadd_pairwise_i8x16_s**
  - `y = i16x8.extadd_pairwise_i8x16_s(x)` is lowered to `SADDLP Vy.8H, Vx.16B`
- **i16x8.extadd_pairwise_i8x16_u**
  - `y = i16x8.extadd_pairwise_i8x16_u(x)` is lowered to `UADDLP Vy.8H, Vx.16B`
- **i32x4.extadd_pairwise_i16x8_s**
  - `y = i32x4.extadd_pairwise_i16x8_s(x)` is lowered to `SADDLP Vy.4S, Vx.8H`
- **i32x4.extadd_pairwise_i16x8_u**
  - `y = i32x4.extadd_pairwise_i16x8_u(x)` is lowered to `UADDLP Vy.4S, Vx.8H`

ARMv7 processors with NEON instruction set
--------------------------------------------------

- **i16x8.extadd_pairwise_i8x16_s**
  - `y = i16x8.extadd_pairwise_i8x16_s(x)` is lowered to `VPADDL.S8 Qy, Qx`
- **i16x8.extadd_pairwise_i8x16_u**
  - `y = i16x8.extadd_pairwise_i8x16_u(x)` is lowered to `VPADDL.U8 Qy, Qx`
- **i32x4.extadd_pairwise_i16x8_s**
  - `y = i32x4.extadd_pairwise_i16x8_s(x)` is lowered to `VPADDL.S16 Qy, Qx`
- **i32x4.extadd_pairwise_i16x8_u**
  - `y = i32x4.extadd_pairwise_i16x8_u(x)` is lowered to `VPADDL.U16 Qy, Qx`

### CODE REVIEW

**Comment by omnisip** [comment]
@Maratyszcza You beat me to it! :-)  I was just about finished my proposal on this.  ;-)

**Comment by omnisip** [comment]
@Maratyszcza 
This might be a little bit more performant for the i16x8->i32x4 variant.  It's also going to work on SSE2. 

Unsigned:
```assembly
        movdqa  xmm1, [wasm_i32x4_splat(0x0000ffff)]
        pand    xmm1, xmm0
        psrld   xmm0, 16
        paddd   xmm0, xmm1
```

Signed:
```assembly
        movdqa  xmm1, xmm0
        pslld   xmm1, 16
        psrad   xmm1, 16
        psrad   xmm0, 16
        paddd   xmm0, xmm1
```

**Comment by Maratyszcza** [comment]
@omnisip You snippets destroy the input `x` value, WAsm engines are not allowed to do this in general.

**Comment by omnisip** [comment]
> @omnisip You snippets destroy the input `x` value, WAsm engines are not allowed to do this in general.

Got it. I corrected the examples above and added a movdqa to prevent clobbering an input register.

**Comment by ngzhian** [comment]
https://crrev.com/c/2513872 prototypes this for arm64.

File: proposals/simd/SIMD.md [hunk 1]
```diff
@@ -411,6 +411,29 @@ def S.neg(a):
     return S.lanewise_unary(neg, a)
 ```
 
+## Extended integer arithmetic
+
+### Extended integer multiplication
```
Comment (tlively) [inline proposals/simd/SIMD.md:416]:
```suggestion
### Extended pairwise integer addition
```

**Comment by Maratyszcza** [comment]
I evaluated this proposal on the O(N^3) part of the 3-Point Correlation computation, described [here](https://github.com/WebAssembly/simd/pull/379#issuecomment-760413099). The use of extended pairwise addition instructions is guarded by `USE_EXTPADD` macro. Results on ARM64 systems are presented below:

Implementation | `USE_EXTPADD=0` | `USE_EXTPADD=1` | Speedup |
-- | -- | -- | --
Snapdragon 670 (Pixel 3a) | 188 us | 164 us | 15% |
Exynos 8895 (Galaxy S8) | 125 us | 114 us | 10% |

As evidenced by the above results, even though the algorithm is not particularly rich on pairwise summations, extended pairwise addition instructions bring noticeable speedup.

**Comment by Maratyszcza** [comment]
Results on x86-64 systems are presented below:

Implementation | `USE_EXTPADD=0` | `USE_EXTPADD=1` | Speedup |
-- | -- | -- | --
AMD Pro A10-8700B | 93 us | 93 us | 0% |
AMD A4-7210 | 303 us | 300 us | 1% |
Intel Xeon W-2135 | 66 us | 66 us | 0% |
Intel Celeron N3060 | 396 us | 396 us | 0% |

Unlike ARM64, x86-64 results demonstrate mostly unchanged performance. However, the x86-64 performance is hindered by suboptimal loading of constants in V8, and could be expected to improve in the future.

**Comment by Maratyszcza** [comment]
@penzn I disagree that N-point correlation is not a suitable workload for evaluation, but nonetheless did another experiment within XNNPACK to evaluate these instruction. In the experiment I added two new variants of 8-bit fixed-point GEMM/IGEMM microkernels: one using Extended Multiplication instructions in combination with the proposed Extended Pairwise Addition instructions, and another using the Extended Multiplication instructions in combination with the widening and addition instructions (not equivalent to extended pairwise addition in general, but works in this particular case and is cheaper than emulation of the extended pairwise addition operation). In these microkernels the extended pairwise addition instructions are the hot loop and their effect on end-to-end performance is more pronounced than in the N-point correlation kernels. Results for end-to-end inference performance are presented in the table below:

Processor (Device) | NN Model | Time with widen+add | Time with extended pairwise addition | Speedup |
-- | -- | -- | -- | --
AMD PRO A10-8700B | MobileNet v1 | 129 ms | 94 ms | 37% |
AMD PRO A10-8700B | MobileNet v2 | 80 ms | 61 ms | 31% |
AMD A4-7210 | MobileNet v1 | 319 ms | 280 ms | 14% |
AMD A4-7210 | MobileNet v2 | 200 ms | 180 ms | 11% |
Intel Xeon W-2135 | MobileNet v1 | 91 ms | 51 ms | 78% |
Intel Xeon W-2135 | MobileNet v2 | 55 ms | 34 ms | 61% |
Intel Celeron N3060 | MobileNet v1 | 414 ms | 436 ms | -5% |
Intel Celeron N3060 | MobileNet v2 | 262 ms | 272 ms | -4% |
Qualcomm Snapdragon 670 (Pixel 3a) | MobileNet v1 | 203 ms | 137 ms | 48% |
Qualcomm Snapdragon 670 (Pixel 3a) | MobileNet v2 | 135 ms | 101 ms | 34% |
Exynos 8895 (Galaxy S8) | MobileNet v1 | 224 ms | 116 ms | 93% |
Exynos 8895 (Galaxy S8) | MobileNet v2 | 145 ms | 89 ms | 63% |

**Comment by Maratyszcza** [comment]
The source code changes for the last experiment can be seen [here](https://github.com/google/XNNPACK/pull/1271)

**Comment by penzn** [comment]
> @penzn I disagree that N-point correlation is not a suitable workload for evaluation

I don't think I ever said that - it is a fine test, what I meant is that we were using something different from use cases listed in PR description. XNNPACK is a much better example - thank you for that!

**Comment by dtig** [comment]
Adding a preliminary vote for the inclusion of extended pairwise addition operations to the SIMD proposal below. Please vote with -

👍 For including extended pairwise addition operations
👎 Against including extended pairwise addition operations

**Comment by fbarchard** [comment]
In native ARM I've used paddl, padal alot.  They've very fast and useful.
On Intel I've used the trick of vpmaddubsw with constants of 1 to achieve the same.  The instruction is very fast, even with the redundent multiply.
Ideally we'd have the accumulate version, but paddl is still useful, followed by an add.
An example of using it for an image is a 2x2 down sample.  paddl to add values horizontally in pairs.  Then add the results, and rounding shift by 2 to get average of the 4 pixels.
It can also be used to accumulate values, such as popcnts.

**Comment by penzn** [comment]
I am not 100% convinced - there are drops in performance on some platforms, plus this is, again, short microkernels.

**Comment by Maratyszcza** [comment]
> plus this is, again, short microkernels.

The XNNPACK results are end-to-end

**Comment by penzn** [comment]
> The XNNPACK results are end-to-end

@Maratyszcza sorry, "microkernels" threw me off, you did mention they were end to end.

**Comment by Maratyszcza** [comment]
Updated end-to-end results in XNNPACK to reflect [the recent optimizations in V8](https://chromium-review.googlesource.com/c/v8/v8/+/2643242).

**Comment by ngzhian** [comment]
The XNNPACK benchmark only use `__builtin_wasm_extadd_pairwise_i16x8_s_i32x4`, I think we can probably expect similar (good) speedups for the other 3 instructions. I see examples of the other 3 instructions in the listed use cases too.

Side note: the naming of this intrinsic seems off? Should it be `__builtin_wasm_extadd_pairwise_i32x4_i16x8_s` instead? (dst shape first).

**Comment by Maratyszcza** [comment]
> The XNNPACK benchmark only use `__builtin_wasm_extadd_pairwise_i16x8_s_i32x4`, I think we can probably expect similar (good) speedups for the other 3 instructions. I see examples of the other 3 instructions in the listed use cases too.

The 3-Point Correlation experiment used `__builtin_wasm_extadd_pairwise_i8x16_u_i16x8` and `__builtin_wasm_extadd_pairwise_i16x8_u_i32x4`.

> Side note: the naming of this intrinsic seems off? Should it be `__builtin_wasm_extadd_pairwise_i32x4_i16x8_s` instead? (dst shape first).
cc @tlively

**Comment by tlively** [comment]
Don't worry about that name. The `__builtin_wasm_*` functions are not intended to be user-facing and follow their own weird naming conventions. We will add properly named user-facing intrinsics eventually if we approve these instructions.

**Pull request merged**